### PR TITLE
[microTVM][Zephyr] Add 'config_main_stack_size' option to API server

### PR DESCRIPTION
**Pull request by mehrdadh**
- When using API server in for external project, sometime we need to set main stack size for zephyr board if model is large. This PR adds this option to Zephyr project API.
- In addition, this PR moves zephyr board properties to a json file to create a single source for board information for testing purposes.
- Finally, this PR adds a validation check for projectOption passed to api server.

cc @areusch @gromero 

waiting for https://github.com/apache/tvm/pull/9018 before merging this.

### CODE REVIEW

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (areusch) [inline apps/microtvm/zephyr/template_project/microtvm_api_server.py:64]: can you compute path relative to API_SERVER_DIR?

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (areusch) [inline apps/microtvm/zephyr/template_project/microtvm_api_server.py:268]: can we say main_stack_size_bytes or config_main_stack_size?

File: tests/micro/zephyr/test_zephyr.py
Comment (areusch) [inline tests/micro/zephyr/test_zephyr.py:61]: can you use same variable?

File: tests/micro/zephyr/conftest.py
Comment (areusch) [inline tests/micro/zephyr/conftest.py:59]: can we place test utils in a test_utils.py instead of conftest?

**Review by areusch — CHANGES_REQUESTED** [review]
thanks @mehrdadh

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (gromero) [inline apps/microtvm/zephyr/template_project/microtvm_api_server.py:69]: A couple of suggestions here.

I think it's not necessary to enforce `BOARD_PROPERTIES = None` before `with` since it won't create a new scope -- please experiment with that possibility.

I prefer the long form, e.g. 
```
with open(BOARDS) as boards:
    BOARD_PROPERTIES = json.load(boards)
```
I do expect the linter won't complain about it.

I think `with` should be inside a `try` in case the file is missing, corrupt etc for whatever reason, and `except` so could raise a BoardProperties(exception) class like there is already one for `BoardError` and `BoardAutodetectFailed`. I recall @areusch concerned once about overloading too much exceptions in the server side, so I'm wondering what he thinks about that suggestion actually :)

File: tests/micro/zephyr/test_zephyr.py
Comment (gromero) [inline tests/micro/zephyr/test_zephyr.py:74]: I think this is the first time an option is passed that is of type 'None'. So far usually what we do if an option is deemed `None` is to simply don't pass the option and then use the following in the server side:
```
if `options.get("main_stack_size"):
    ... options["main_stack_size"]
```
So I personally prefer to keep this way instead of passing an option with value "None".  If that form is adopted in this case it would be a matter of checking for stack_size after it's set accordingly and then check if it's None, appending (or not) the "main_stack_size" option to the options dict. @areusch wdyt?

File: python/tvm/micro/project.py
Comment (gromero) [inline python/tvm/micro/project.py:104]: I'm wondering if it would be possible and better to rewrite that as:

```
available_options = [option["name"] for option in template.info()["project_options"]]
if not set(options.keys()).issubset(available_options):
   raise ValueError(...)
```

File: python/tvm/micro/project.py
Comment (gromero) [inline python/tvm/micro/project.py:111]: s/valie/valid/ if that form will be kept?

**Review by gromero — CHANGES_REQUESTED** [review]
Hi @mehrdadh Thanks for the change. Looks good overall. 

Please see my comments inline. And a nit in the title: I think it should be:
` [microTVM][Zephyr] Add 'main_stack_size' option to API server`

Cheers

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (mehrdadh) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py:64]: done.

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (mehrdadh) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py:69]: since this is used once I think we could just use FileNotFoundError exception? something like this:

```
try:
    with open(BOARDS) as boards:
        BOARD_PROPERTIES = json.load(boards)
except FileNotFoundError:
    raise FileNotFoundError(f"Board file {{{BOARDS}}} does not exist.")
```

File: tests/micro/zephyr/conftest.py
Comment (mehrdadh) [reply tests/micro/zephyr/conftest.py:59]: done.

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (gromero) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py:69]: @mehrdadh yeah I think that's fine too. Btw, do we need triple `{` `}` here?

File: tests/micro/zephyr/test_zephyr.py
Comment (mehrdadh) [reply tests/micro/zephyr/test_zephyr.py:74]: I agree. changed it.

File: python/tvm/micro/project.py
Comment (mehrdadh) [reply python/tvm/micro/project.py:104]: done.

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (mehrdadh) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py:69]: yeah because of f"" statement

**Comment by mehrdadh** [comment]
@gromero thanks for the review. I addressed the comments. PTAL.

**Review by gromero — CHANGES_REQUESTED** [review]
@mehrdadh Thanks for addressing the comments. There is only one remaining issue in the code that was not caught by the CI (PTAL at the inline comments) and a nit in the PR title. Otherwise LGTM.

File: tests/micro/zephyr/test_zephyr.py
Comment (mehrdadh) [reply tests/micro/zephyr/test_zephyr.py:53]: thanks for catching this. Done.

**Comment by mehrdadh** [comment]
waiting for CI fix: https://github.com/apache/tvm/pull/9050

File: apps/microtvm/zephyr/template_project/microtvm_api_server.py
Comment (gromero) [reply apps/microtvm/zephyr/template_project/microtvm_api_server.py]: @mehrdadh ah, I see, you want to have the file name between `{` and `}` too. OK.

**Comment by gromero** [comment]
@mehrdadh LGTM, thanks! Once CI is happy with it I'm happy too.

PS: that branch is still failing on disco board on `test_autotune_conv2d` but the error is not related to the changes in this PR. It seems the error was introduced with AutoTVM patchset. Anyway, I've filed https://github.com/apache/tvm/issues/9049 to keep track of it.

**Review by gromero — APPROVED** [review]

**Review by areusch — APPROVED** [review]
@mehrdadh please fix the CI issue

**Comment by areusch** [comment]
blocked on https://github.com/apache/tvm/issues/9013

**Pull request merged**
